### PR TITLE
[inductor] Rewrite convolution triton templates

### LINKED ISSUE #95775: Compatibility issue of torch._inductor.triton_ops.conv_perf_model and latest nightly triton
### 🐛 Describe the bug

In <https://github.com/openai/triton/commit/20100a7254e62efd0fced864b52a877c520fc38a#diff-e15e164b247309ce1b9a886812ae86b7043b4350af85f1b7cffe9a1f4dc45c05L361> (`python/src/triton.cc`, `init_triton_runtime()`), triton removed many attributes of `triton._C.libtriton.triton.runtime` such as `cc`, `max_shared_memory`, which is called in `triton_ops/conv_perf_model` in inductor. This issue can be triggered by setting `torch._inductor.config.triton.convolution` to `triton` and compiling a model with conv.

Pinning triton to a version prior to this change (such as `2.0.0.dev20221030`) workarounds this issue.

### Error logs

```
Traceback (most recent call last):
  File "/home/username/pytorch2/test3.py", line 34, in <module>
    print(model(torch.randn(1, 3, 224, 224).cuda()))
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 82, in __call__
    return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/username/pytorch2/test3.py", line 11, in forward
    def forward(self, x):
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2821, in forward
    return compiled_fn(full_args)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1222, in g
    return f(*args)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2388, in debug_compiled_function
    return compiled_function(*args)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1900, in runtime_wrapper
    all_outs = call_func_with_args(
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1247, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1222, in g
    return f(*args)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 2153, in forward
    fw_outs = call_func_with_args(
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py", line 1247, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_inductor/compile_fx.py", line 250, in run
    return model(new_inputs)
  File "/tmp/torchinductor_username/xf/cxfscyw26zjrfhp4tamvn7hmhsnr6qtiq232xccma7l6sdo3u7dr.py", line 90, in call
    buf2 = triton_ops.conv(buf1, primals_1, None, (1, 1), (0, 0), (1, 1), False, (0, 0), 1)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_inductor/triton_ops/conv.py", line 732, in forward
    return _conv._call(
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_inductor/triton_ops/conv.py", line 606, in _call
    _kernel_delta_x[grid](
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/triton/runtime/autotuner.py", line 75, in run
    pruned_configs = self.prune_configs(kwargs)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/triton/runtime/autotuner.py", line 95, in prune_configs
    pruned_configs = self.early_config_prune(self.configs, self.nargs)
  File "/staff/username/miniconda3/envs/pytorch2/lib/python3.10/site-packages/torch/_inductor/triton_ops/conv_perf_model.py", line 100, in early_config_prune
    cc = _triton.runtime.cc(backend, device)
AttributeError: module 'triton._C.libtriton.triton.runtime' has no attribute 'cc'
```

### Minified repro

```python
# it seems that minifier failed to generate a minified code with `TORCHDYNAMO_REPRO_AFTER="aot"`
# so I will just put my MRE here.
import torch
import torch._dynamo
import torch._inductor.config
import logging

class ModuleTest2(torch.nn.Module):
    def __init__(self, *args, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self.weight = torch.nn.Parameter(torch.randn(1, 3, 224, 224))
    
    def forward(self, x):
        x = x + torch.ones_like(x)
        res = torch.nn.functional.conv2d(x, self.weight)
        res = res + torch.ones_like(res)
        return res

torch._dynamo.config.log_level = logging.DEBUG
torch._dynamo.config.output_code = True
torch._dynamo.config.verbose = True
torch._inductor.config.triton.convolution = "triton"
torch._inductor.config.debug = True

model = ModuleTest2().cuda()

model = torch.compile(model)
print(model)
print(model(torch.randn(1, 3, 224, 224).cuda()))
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0.dev20230228
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 15.0.1 (https://github.com/exaloop/llvm-project 55b0b8fa1c9f9082b535628fc9fa6313280c0b9a)
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.10.9 (main, Jan 11 2023, 15:21:40) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-137-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A100-PCIE-40GB
Nvidia driver version: 465.19.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              32
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
Stepping:                        6
Frequency boost:                 enabled
CPU MHz:                         900.000
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        5200.00
Virtualization:                  VT-x
L1d cache:                       3 MiB
L1i cache:                       2 MiB
L2 cache:                        80 MiB
L3 cache:                        96 MiB
NUMA node0 CPU(s):               0-31,64-95
NUMA node1 CPU(s):               32-63,96-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0.dev20230228
[pip3] torchaudio==2.0.0.dev20230301
[pip3] torchvision==0.15.0.dev20230228
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pytorch                   2.0.0.dev20230228 py3.10_cuda11.8_cudnn8.7.0_0    pytorch-nightly
[conda] pytorch-cuda              11.8                 h7e8668a_3    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                2.0.0.dev20230301     py310_cu118    pytorch-nightly
[conda] torchvision               0.15.0.dev20230228     py310_cu118    pytorch-nightly
```

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95558
* __->__ #95556

Fixes #95775

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: benchmarks/dynamo/microbenchmarks/operatorbench.py [hunk 1]
```diff
@@ -132,7 +131,7 @@ def skip_operator(operator):
         print(f"Skipping {operator}, no inductor impl")
         return True
 
-    if inductor_config.triton.convolution == "aten" and "convolution" in str(operator):
+    if "convolution" in str(operator):
```
Comment (chuanqi129) [inline benchmarks/dynamo/microbenchmarks/operatorbench.py:134]: Hi @jansel , the changes of this opbenchmark scripts also need for [release 2.0](https://github.com/pytorch/pytorch/blob/release/2.0/benchmarks/dynamo/microbenchmarks/operatorbench.py#LL135), do we need to add this change into release 2.0 branch also?

Comment (chuanqi129) [reply benchmarks/dynamo/microbenchmarks/operatorbench.py:134]: Yes, you're right, it will not be included in binaries.

**Comment by jansel** [comment]
@Chillee or @ngimel mind reviewing this one?

File: torch/_inductor/config.py [hunk 2]
```diff
@@ -37,7 +37,7 @@
 benchmark_harness = True
 
 # fuse pointwise into templates
-epilogue_fusion = False
+epilogue_fusion = True
```
Comment (Chillee) [inline torch/_inductor/config.py:40]: Is this intentional?

File: torch/_inductor/kernel/conv.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    )
+
+
+def channels_last_order(rank):
+    order = list(reversed(range(rank)))
+    order.insert(1, order.pop(-1))
+    return order
+
+
+def convert_1x1_conv_to_mm(x, weight, bias):
```
Comment (Chillee) [inline torch/_inductor/kernel/conv.py:229]: Didn't check this closely but seems reasonable.

File: torch/_inductor/kernel/conv.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    if bias is not None and ir.get_device_type(x) != "cpu":
+        # peel off the bias, cudnn is slower with it
+        result = convolution(x, weight, None, **kwargs)
+        return L[aten.add](
+            result, L[aten.view](bias, [result.get_size()[1]] + ndim * [1])
+        )
+
+    x.realize()
+    weight.realize()
+    layout = conv_layout(x, weight, None, **kwargs)
```
Comment (Chillee) [inline torch/_inductor/kernel/conv.py:317]: So did we remove the option for "tune_layout"?

**Review by Chillee — APPROVED** [review]
Nice -2700 lines :)

File: torch/_inductor/config.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/config.py:40]: Yes

File: torch/_inductor/kernel/conv.py (hunk 4 above)
Comment (jansel) [reply torch/_inductor/kernel/conv.py:317]: Yeah, will need to re-add it in another PR.

File: torch/_inductor/kernel/conv.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        # there are some odd models where this check fails (e.g. shufflenet_v2_x1_0)
+        and V.graph.sizevars.maybe_guard_equals(in_chan, x.get_size()[1])
+    ):
+        if (
+            is_ones(kernel_shape)
+            and is_ones(stride)
+            and is_zeros(padding)
+            and groups == 1
+        ):
+            choices.append(aten_conv1x1_via_mm.bind(args, layout))
```
Comment (ngimel) [inline torch/_inductor/kernel/conv.py:348]: where is bias addition handled for this one (and for triton templates below)?

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/kernel/conv.py (hunk 5 above)
Comment (jansel) [reply torch/_inductor/kernel/conv.py:348]: We split off the bias into another op on line 312.

Comment (ngimel) [reply torch/_inductor/kernel/conv.py:348]: For cuda, shouldn't we add addmm to 1x1 options with bias?

Comment (jansel) [reply torch/_inductor/kernel/conv.py:348]: Yeah, I think that would be a good option to add.  I also want to add alternate layout options.

**Pull request closed without merging**